### ISSUE TITLE
Add Scheduled Publish functionality

### ISSUE THREAD

**Issue by johnrom** [OrchardCMS/OrchardCore#3991]
Orchard 1 "Publish Later" documentation: https://docs.orchardproject.net/en/latest/Documentation/Saving-scheduling-and-publishing-drafts/

Wondering how we might implement this important content management facility. Should it be in Core? Should it be a module?

- There is already `bool ContentItem.Published` and `DateTime ContentItem.PublishedUtc`. It might make sense to leverage this existing PublishedUtc DateTime in any scheduling functionality.

### Options

- Use a Module: 
    - Module cannot easily override IsPublished() content extension, so this might make things more confusing. 
    - Maybe the IsPublished() content extension also checks against PublishedUtc, but by default publishing in Core sets this to the current time, which can be overridden by a module.
- Implement in Core:
    - Add `bool Schedulable` to `ContentTypeSettings`.

**Comment by sebastienros** [comment]
I suggest to do a custom part, that would be added to the content types we want to be able to schedule. And it would use another module that manages scheduled tasks.

It might be related to this issue https://github.com/OrchardCMS/OrchardCore/issues/3187

**Comment by johnrom** [comment]
@sebastienros does that it would or wouldn't use `PublishedUtc`? if it doesn't then I'm not sure what the point of PublishedUtc would be, since I think most sorts would want to use the scheduled publish time.

**Comment by sebastienros** [comment]
PublishedUtc is the time when it "was" last published. CreatedUtc is the date it was created or first published. A scheduled publishing task module would call `ContentMAnager.PublishAsync()` on this content item which would itself update the `PublisedUtc` value. I suggest you look at O1.

**Comment by sebastienros** [comment]
`ITaskSchedulerProvider` which has `GetNextTasks(Timespan duration)` to return tasks that should be processed in the next duration time. There will be multiple implementations. One will store scheduled tasks for specific times, like the PublishLater or ArchiveLater. Another one can handle recurring tasks, based on patterns like "every day", but this will be done later.

`IScheduledTaskManager` that provides methods to Load and Save a scheduled task. This would use a single yessql document by default, and a change token to notify when something has changed.

`ITaskSchedulerManager` that can store the log of previously run scheduled tasks. Including output from a task, like success, failure, or messages.

and `IScheduledTaskBackgroundJob` that runs every minute to process scheduled tasks. It has a list of scheduled tasks that is computed by looking at all scheduled tasks, and creating a list of next events that will occur in the next "hour" for instance. And this list is cached for less than one hour, so asking for the next task is mostly always free. It gets its list from all  `ITaskSchedulerProvider` implementations.

This list can contain tasks that are in the past if it ran into an issue. For instance a content item that was scheduled to be published earlier and it failed. We need to be able to deduplicate scheduled tasks too, by a logical identifer, such that a recurring task that gets duplicated is run only once. For instance a LuceneIndex task would have a string making it unique such that only one is required.

/cc @jtkech I'd like to check with you if we haven't designed something related to this with background tasks, that is overlapping with these requirements.

**Comment by jtkech** [comment]
Yes, i remember we talked about it, i don't remember all details but i will think about it and follow this issue. Yes, i think it should be tied to how background tasks are working through the background servive, and we can already see how `IndexingTaskManager`, `CreateIndexingTaskContentHandler` and the lucene `IndexingBackgroundTask` are working, but here with a specific `IndexingTask` table.

As a side note, i'm also working on #4733 based on the @Skrypt lucene branch #3809 using a new  `IndexLatest` option, we did some work on `LuceneIndexingContentHandler`, but need to be consistent with what we are doing on the indexes through the admin or the `IndexingBackgroundTask`.

E.g i saw that maybe we would have to change the `IndexingTask` table format to add a `ContentType` info because an index may not include all content types, but i'm still specific here. So, maybe we would have to find a more global format for all these kind of tasks, and maybe with an additional column in json format holding data for specific usages.

**Comment by sebastienros** [comment]
Look at the content of this table and then you can correct yourself ;)

**Comment by jtkech** [comment]
I already did it, maybe you mean the `ContentItemId` column, so when we grab this item we will know its `ContentType`, but i was thinking about not querying an item if its type doesn't belong to any index.

But i just quickly saw it yesterday while working on the @Skrypt lucene branch, need to see it more in details. The stuff already done in the dev branch for lucene and indexing needs to be consistent with the new `IndexLast` option and the fact that an index may not include all types.

Maybe this is just the indexing handler that needs to not enlist a indexing task for an item whose type doesn't belong to any index in place of having to check it in `ProcessContentItemsAsync()`. But an index settings may change afterwards and `CreateIndexingTaskContentHandler` is not aware of lucene.

But no problem, i will see it more in details, maybe not a problem because we use batch queries, so if anyway an item is grabbed we can then easily check its content type.

**Comment by Skrypt** [comment]
Also related to @jtkech comment, in my PR we still need to distinct wether we want to index Latest or Published records as I broke this feature by adding the perf fixes. Though, as we seen last night it might only require to simply do the proper _contentManager.GetAsync() method call. By default it will return only published content items.

**Comment by sebastienros** [comment]
Lucene has nothing to do with this issue. It's a different kind of task, not tied to any DateTime. These are just markers for each content item that they have to be indexed anytime an index needs it. And the settings for content types, or publication state have nothing to do with it.

**Comment by Skrypt** [comment]
Oh but I think so. Because if you want to publish later then if the contentitemId is marked to be indexed by Lucene it will index it and return it as a search result even if the content item should be published later on.

**Comment by sebastienros** [comment]
If the indexer only indexes published items, it will only be when the item is published. My comment was more for @jtkech to understand the key differences in the "tasks".

**Comment by Skrypt** [comment]
Yeah but in my PR we allow indexing also drafts so in @jtkech PR there are some conditions to validate what we are doing with the indexing.

**Comment by jtkech** [comment]
> to understand the key differences in the "tasks".

Yes i know, i just thought about generalizing some things seing some analogies with the indexing tasks, and because in your comment you was also talking about `LuceneIndex task` ;)

**Comment by johnrom** [comment]
I've looked through a bit of the code, and I'll draw up a simple (😆) outline of how it looks like it should work. I've taken the liberty of naming things a certain way, but won't be offended if they are changed.

- Currently, the background task processor is `ModularBackgroundService`, an `IHostedService` running in the background.
- Every minute, it iterates through each tenant, grabs each registered tenant `IBackgroundTask` through DI, puts those which need to run into the processing queue, and runs them.
- The Lucene `IndexingBackgroundTask` and others register themselves as Cron-scheduled tasks just by inheriting IBackgroundTask with a `[BackgroundTask]` attribute. 
- The `ScheduledTaskProcessingBackgroundTask` (name?!) would be a new implementation of this `IBackgroundTask`, which registers itself minutely on Cron. It gathers available tasks from `IScheduledTasksManager.GetUpcomingTasks(TimeSpan duration)` once an hour (ish) or when tasks have changed based on a changeToken provided to IScheduledTasksManager.
- The `IScheduledTasksManager` will manage adding, removing, and modifying raw `ScheduledTask`s in the `ScheduledTasks` Document, signaling the ChangeToken when needed.
- Each `ScheduledTask` will be a ContentItem (similar to how CustomSettings are really ContentItems) which has a type of `ScheduledTaskRecord` consumable by `IScheduledTasksProvider`, for example `RecurringTasksProvider`, `ScheduledTasksProvider` and other future implementations.
- `ScheduledTasksProvider.GetTasksToRun(TimeSpan duration)` will check if each `ScheduledTask` has a given type, convert to a `ScheduledTaskSettings`, check if its time has come, and other conditions, and if so, it will return the `ScheduledTask` with a specific time that it will run.
- `ScheduledTaskProcessingBackgroundTask` will save this list ordered by the specific time that it should run, and this and future runs will execute a batch of tasks that should have executed already. It creates a ScheduledTaskContext just like O1 and runs all registered `IScheduledTaskHandler`s against it.
- `PublishLaterTaskHandler` checks that the task is of type Publish (or whatever), and if so, runs contentItem.Publish.

When a ContentItem with PublishLaterPart is being "Published Later", it actually saves a Draft, clears any existing Publish Later tasks scheduled for this content item, and adds a new one for the specified DateTime.

**Comment by pbros** [comment]
Hello, any progress on this functionality? It would be useful for us as well!

**Comment by Piedone** [comment]
We'd potentially work on this.

I don't think this requires (or would benefit from too much) from a scheduled task infrastructure. How I'd do it is as following, please comment:

- New content PublishLater (or ScheduledPublish) content part as discussed above. This would contain the future publish datetime in an indexed field.
- A background task running every minute fetches all items where scheduled datetime < current datetime, and publishes these.

Am I missing something? The only issue I see is that in a multi-node environment this process should run on a single node only (but we'd need such locking for background tasks any way).

**Comment by johnrom** [comment]
@piedone this is a task scheduled minutely. the point that @sebastienros was making that if we're going to create a scheduled task, we should create a unified infrastructure for creating _any_ scheduled task in one fell swoop. There should only be one mechanism checking for tasks and scheduling them to be run, or the site will be doing extra work in the background, and testing becomes more complex.

My thoughts above were meant to be reviewed by a core maintainer to guide implementation, so if they haven't signed off it may not be quite as complex as I imagined it to be, so I'd definitely wait for feedback before starting. If I get to the point where my team needs this, I was going to tackle it based off of the above -- but I haven't gotten there and may not.

**Comment by Piedone** [comment]
Yeah, though due to how simple this can work as I explained above I don't think building scheduled tasks with this would actually be a proper use-case (it needs background tasks all right, but we have those already so it wouldn't contain a parallel piece of infrastructure).

**Comment by Piedone** [comment]
We'll start working on the Publish Later module but not on the scheduled task management infrastructure.

**Comment by johnrom** [comment]
A minutely background task is way more expensive than scheduling a single task at the exact time that a document is set to be published in the future, and checking for tasks instead of checking all posts minutely. Especially if there are thousands of posts. O(n)/min where n = documents vs O(n)/min where n = scheduled tasks. If scalability is planned for enterprise level or news publications, it'll be a big difference over time.

**Comment by Piedone** [comment]
There will be a single query running once every minute to fetch all content items where the scheduled publish date has passed (so it's not one query for every content item). In theory, this is indeed less efficient than a centralized task manager checking a scheduled task table every minute, in the case when there are multiple such background-checking modules working at the same time (like Publish Later and Archive Later). However, at the time there is no other such module.

In the future, this feature will be possible to refactor to use a then available scheduled task service.

**Comment by Piedone** [comment]
Is it OK for everyone if I create a new issue about scheduled tasks and Archive Later?

**Comment by jxt-seth** [comment]
Do you think you will have to bring in some sort of content status extensions to deliver this? 

At the moment there is DRAFT and PUBLISHED but I feel there is a need for more status information.

For example, currently, if you publish an item then you edit it and save a draft it just says "draft". OC  doesn't inform you anywhere that a version of the item has actually been published and is currently live BUT there is also a draft version as well - it just says "draft".

Almost needs more clarification:

- **Draft** (when the item has never been published)
- **Published** (when the item is published and there is no draft version currently)
- **Draft, newer than published** (when the item is published and there is also a draft version newer than published version)
- **Scheduled publish @ X** (draft item is scheduled to publish. it has not been published before)
- **Scheduled publish @ X, newer than published** (item is published and there is also a draft version scheduled to replace the published version)
- **Published, scheduled unpublish @ X** (item is published and is scheduled to unpublish)
- **Unpublished** (item has been unpublished, either manually or via a scheduled upnpublish, and no further edits have been made since that time. If you were to edit and save a draft it would then go back to "draft" status)

I know this is more that "Scheduled publish" functionality but if you are going to add scheduled publishing then you need to be showing more info about the state of the item IMHO.

**Comment by Piedone** [comment]
This is an interesting idea but only partially builds what's established now with the module. So please create a new issue about this so we can close this with the current scope, let people use it and go from there with new ideas.

**Comment by Piedone** [comment]
I've created two new issues for Archive Later and scheduled tasks otherwise this will be never-ending feature creep :).

**Comment by Piedone** [comment]
This is already released in RC2.

**Comment by johnrom** [comment]
@piedone finally circling back to try this out on another project, after requesting it so many years ago. works fantastic. big thanks!

**Comment by Piedone** [comment]
I'm glad to hear, thanks for letting me know!
